app/coffee/DocumentUpdaterController.coffee

```diff
@@ -59,30 +59,27 @@ module.exports = DocumentUpdaterController =
 				HealthCheckManager.check channel, message.key
 
 	_applyUpdateFromDocumentUpdater: (io, doc_id, update) ->
-		clientList = io.sockets.clients(doc_id)
-		# avoid unnecessary work if no clients are connected
-		if clientList.length is 0
-			return
-		# send updates to clients
-		logger.log doc_id: doc_id, version: update.v, source: update.meta?.source, socketIoClients: (client.id for client in clientList), "distributing updates to clients"
-		seen = {}
-		# send messages only to unique clients (due to duplicate entries in io.sockets.clients)
-		for client in clientList when not seen[client.id]
-			seen[client.id] = true
-			if client.id == update.meta.source
-				logger.log doc_id: doc_id, version: update.v, source: update.meta?.source, "distributing update to sender"
-				client.emit "otUpdateApplied", v: update.v, doc: update.doc
-			else if !update.dup # Duplicate ops should just be sent back to sending client for acknowledgement
-				logger.log doc_id: doc_id, version: update.v, source: update.meta?.source, client_id: client.id, "distributing update to collaborator"
-				client.emit "otUpdateApplied", update
-		if Object.keys(seen).length < clientList.length
-			metrics.inc "socket-io.duplicate-clients", 0.1
-			logger.log doc_id: doc_id, socketIoClients: (client.id for client in clientList), "discarded duplicate clients"
+		source = update.meta?.source
+		if io.sockets.connected.hasOwnProperty(source)
+			emitter = io.sockets.connected[source]
+			logger.log {doc_id, version: update.v, source}, "distributing update to sender"
+			emitter.emit "otUpdateApplied", v: update.v, doc: update.doc
+		else
+			emitter = io
+		return if update.dup
+		logger.log {doc_id, version: update.v, source}, "distributing updates to clients"
+		emitter.to(doc_id).emit "otUpdateApplied", update
 
 	_processErrorFromDocumentUpdater: (io, doc_id, error, message) ->
-		for client in io.sockets.clients(doc_id)
-			logger.warn err: error, doc_id: doc_id, client_id: client.id, "error from document updater, disconnecting client"
-			client.emit "otUpdateError", error, message
-			client.disconnect()
+		io.to(doc_id).clients (err, clientIds) ->
+			if err?
+				return logger.err {room: doc_id, err}, "failed to get room clients"
+
+			for clientId in clientIds
+				continue unless io.sockets.connected.hasOwnProperty(clientId)
+				client = io.sockets.connected[clientId]
+				logger.warn err: error, doc_id: doc_id, client_id: client.id, "error from document updater, disconnecting client"
+				client.emit "otUpdateError", error, message
+				client.disconnect()
```

Comment:
Using the v2 api I think we can write this whole function as

```
io.to(doc_id).emit("otUpdateError", error, message).disconnect()
```

Perhaps the same approach can be used elsewhere.

Reply:
`.disconnect()` can not be distributed to a room like a message can be broadcasted.

```
Uncaught TypeError: io.to(...).emit(...).disconnect is not a function
```

Reply:
Ah yes, missed that in the logs -- it did disconnect all the clients, but not for the right reason!